src/mixer.rs

```diff
@@ -0,0 +1,428 @@
+use std::{ptr, mem};
+
+use alsa;
+use super::card;
+use super::error::*;
+
+const SELEM_ID_SIZE: usize = 64;
+
+/// wraps [snd_mixer_t](http://www.alsa-project.org/alsa-doc/alsa-lib/group___mixer.html)
+pub struct Mixer(*mut alsa::snd_mixer_t);
+
+impl Mixer {
+    /// Creates a new iterator for a specific card using the cards index and
+    /// using hw name, i.e. `hw:0`
+    pub fn new(c: &card::Card) -> Result<Mixer> {
+        let card = &CString::new(format!("hw:{}", c.get_index())).unwrap();
+        let mut mixer_handle = ptr::null_mut();
+        try!(acheck!(snd_mixer_open(&mut mixer_handle, 0)));
+        try!(acheck!(snd_mixer_attach(mixer_handle, card.as_ptr())));
+        try!(acheck!(snd_mixer_selem_register(mixer_handle, ptr::null_mut(), ptr::null_mut())));
+        try!(acheck!(snd_mixer_load(mixer_handle)));
+        Ok(Mixer(mixer_handle))
+    }
+}
+
+/// Closes mixer and frees used resources
+impl Drop for Mixer {
+    fn drop(&mut self) {
+        unsafe { alsa::snd_mixer_close(self.0) };
+    }
+}
+
+/// Wraps [snd_mixer_selem_id_t](http://www.alsa-project.org/alsa-doc/alsa-lib/group___simple_mixer.html)
+#[derive(Copy, Clone)]
+pub struct Elem<'a>{
+    handle: *mut alsa::snd_mixer_elem_t,
+    mixer: &'a Mixer
+}
+
+impl<'a> Elem<'a> {
+    /// Checks if this element is null by checking if the handle is null
+    pub fn is_null(&self) -> bool {
+        self.handle == 0 as *mut alsa::snd_mixer_elem_t
+    }
+}
+
+
+/// Iterator for all elements of mixer
+#[derive(Copy, Clone)]
+pub struct Iter<'a>{
+    last_handle: *mut alsa::snd_mixer_elem_t,
+    mixer: &'a Mixer
+}
+
+impl<'a> Iter<'a> {
+    pub fn new(m: &'a Mixer) -> Iter<'a> {
+        Iter {
+            last_handle: ptr::null_mut(),
+            mixer: m
+        }
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = Elem<'a>;
+
+    fn next(&mut self) -> Option<Elem<'a>> {
+        let elem = if self.last_handle.is_null() {
+            unsafe { alsa::snd_mixer_first_elem(self.mixer.0) }
+        } else {
+            unsafe { alsa::snd_mixer_elem_next(self.last_handle) }
+        };
+
+        if elem.is_null() {
+            None
+        } else {
+            self.last_handle = elem;
+            Some(Elem { handle: elem, mixer: self.mixer})
+        }
+    }
+
+}
+
+/// Wrapper for `snd_mixer_selem_id_t`, using array of hardcoded length
+// #[derive(Copy, Clone)]
+pub struct SelemId([u8; SELEM_ID_SIZE]);
+
+impl SelemId {
+    /// Creates a new SelemId` of hardcoded size SELEM_ID_SIZE.
+    /// This size is checked against `snd_mixer_selem_id_sizeof`
+    pub fn new(elem: Elem) -> SelemId {
+        // Create empty selem_id and fill from mixer
+        let sid = SelemId::empty();
+        unsafe { alsa::snd_mixer_selem_get_id(elem.handle, sid.as_ptr()) };
+        sid
+    }
+
+    /// Returns an empty (zeroed) SelemId. This id is not a useable id and need to be initialized
+    /// like `SelemId::new()` does
+    pub fn empty() -> SelemId {
```

Comment:
We should not have unusable id's, instead have a function that takes name and index and creates an SelemId from that